sdk/identity/Azure.Identity/src/AuthenticationRecord.cs

```diff
@@ -0,0 +1,181 @@
+            if (stream == null) throw new ArgumentNullException(nameof(stream));
+
+            return DeserializeAsync(stream, false, cancellationToken).EnsureCompleted();
+        }
+
+        /// <summary>
+        /// Deserializes the <see cref="AuthenticationRecord"/> from the specified <see cref="Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="Stream"/> from which the serialized <see cref="AuthenticationRecord"/> will be read.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        public static async Task<AuthenticationRecord> DeserializeAsync(Stream stream, CancellationToken cancellationToken = default)
+        {
+            if (stream == null) throw new ArgumentNullException(nameof(stream));
+
+            return await DeserializeAsync(stream, true, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task SerializeAsync(Stream stream, bool async, CancellationToken cancellationToken)
+        {
+            using (var json = new Utf8JsonWriter(stream))
+            {
+
+                json.WriteStartObject();
+
+                json.WriteString(s_usernamePropertyNameBytes, Username);
+
+                json.WriteString(s_authorityPropertyNameBytes, Authority);
+
+                json.WriteString(s_homeAccountIdPropertyNameBytes, HomeAccountId);
+
+                json.WriteString(s_tenantIdPropertyNameBytes, TenantId);
+
+                json.WriteEndObject();
+
+                if (async)
+                {
+                    await json.FlushAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    json.Flush();
+                }
+            }
+        }
+
+        private static async Task<AuthenticationRecord> DeserializeAsync(Stream stream, bool async, CancellationToken cancellationToken)
+        {
+            var authProfile = new AuthenticationRecord();
+
+            using JsonDocument doc = async ? await JsonDocument.ParseAsync(stream, cancellationToken: cancellationToken).ConfigureAwait(false) : JsonDocument.Parse(stream);
+
+            foreach (JsonProperty prop in doc.RootElement.EnumerateObject())
+            {
+                switch (prop.Name)
+                {
+                    case UsernamePropertyName:
+                        authProfile.Username = prop.Value.GetString();
+                        break;
+                    case AuthorityPropertyName:
+                        authProfile.Authority = prop.Value.GetString();
+                        break;
+                    case HomeAccountIdPropertyName:
+                        authProfile.AccountId = new AccountId(prop.Value.GetString());
+                        break;
+                    case TenantIdPropertyName:
+                        authProfile.TenantId = prop.Value.GetString();
+                        break;
+                }
```

Comment:
What should we do if an unknown property shows up here? i.e. should default case throw? Log?

Reply:
For now I was ignoring it and only saving known fields. This makes the serialization non-roundtripable, which I could foresee being counter intuitive and possibly a problem for future versioning. I'll file an issue to resolve this before GA.